### DISCUSSION TITLE
Serving docs in subpath

### DISCUSSION THREAD

**Discussion by tobihans** [juhaku/utoipa#185] [Q&A] [answered]
Hi! It's my first time using this crate.
I have an actix app where I'm versioning the API. I'd like to serve a doc for each version (like `api/v1/swagger-ui`) but configuring the swagger ui under a sub path points to wrong endpoint. The path to the openapi spec doesn't take the prefix path into account.
The code below show how I organized things.
- `main.rs`
```rust
#[actix_web::main]
async fn main() -> std::io::Result<()> {

    HttpServer::new(|| App::new().service(web::scope("/api/v1").configure(api::v1::config)))
        .bind((dotenv!("HOST"), dotenv!("PORT").parse::<u16>().unwrap_or(5000)))?
        .run()
        .await
}
```
- `api/v1.rs`
```rust
pub fn config(cfg: &mut web::ServiceConfig) {
    let openapi = doc::Doc::openapi();
    cfg
        .service(handlers::hello)
        .service(SwaggerUi::new("/swagger-ui/{_:.*}").url("/api-doc/openapi.json", openapi.clone()));
}
```
- `api/v1/doc.rs`
```rust
use utoipa::{OpenApi};
use super::handlers::*;

#[derive(OpenApi)]
#[openapi(handlers(hello))]
pub struct Doc;
```
You can see in the screenshot below that the path is wrong as it doesn't take account of prefixes if they are:
![image](https://user-images.githubusercontent.com/64963758/174969286-568e8995-e359-42c0-9720-0037b32c4c07.png)
When I manually correct to add the prefix, everything is fine:
![image](https://user-images.githubusercontent.com/64963758/174969407-ae2ec2fb-d6e9-4c94-9ef2-ecaefab0af34.png)

I've been searching a way to configure this behavior but I'm unable to find it. I appreciate any help, maybe doc link, explanation or other. Thank you!

**Comment by tobihans** [comment]
I went through the implementation and apparently there is nowhere to get such information (path prefixes if they exist) and take it into account. I will investigate more on actix side to see if it's possible to extract such information. 

But I found a workaround for the moment. I serve the specification on another endpoint under my prefix and then, specify the full path from root in the SwaggerUI config. As I already have my spec under an endpoint, I'm replacing the one I should provide to the `url` method with an empty one.

```rust
// This config is mounted under `/api/v1` on my actix app
pub fn config(cfg: &mut web::ServiceConfig) {
    cfg.route("/openapi.json", web::get().to(doc::openapi_spec))
        .service(handlers::hello)
        .service(SwaggerUi::new("/swagger-ui/{_:.*}").url(
            "/api/v1/openapi.json",
            OpenApi::new(Default::default(), Paths::new()),
        ));
}
```
where `doc::openapi_spec` is implemented like this:
```rust
pub async fn openapi_spec() -> impl Responder {
    HttpResponse::Ok().json(Doc::openapi())
}
```

**Comment by juhaku** [comment, marked as answer]
Yeah scoping is a tricky business in the `actix-web` and other frameworks since information about scopes is not usually available to the third-party library. `actix-web` is not an exception and `utoipa` does not have access to scopes of the `actix-web`. Take a look at this example for serving multiple docs: https://github.com/juhaku/utoipa/blob/master/examples/actix-web-multiple-api-docs-with-scopes/src/main.rs. And also this same thing is being discussed here as well: https://github.com/juhaku/utoipa/issues/121.
 
Snippet from above example with some explanatory comments
```rust
async fn main() -> Result<(), impl Error> {
    #[derive(OpenApi)]
    #[openapi(handlers(api1::hello1))]
    struct ApiDoc1;

    // Closure for HttpServer::new(...) is being called once for each worker 
    // thread of the application. (typically number of cpu cores) 
    // this means that if OpenApi has some modifications during runtime then to get 
    // always same instance of OpenApi doc is to call ApiDoc1::openapi() before the closure and 
    // clone it. (not the way its done here)
    // This also is valid for with any third-party initialization e.g. databases etc.
    HttpServer::new(move || { 
        App::new()
            .service(
                web::scope("/api") // <- this scope is not known to `utoipa` OpenApi doc.
                    .service(api1::hello1),
            )
            // the /swagger-ui url is relative from the root of the server
            .service(SwaggerUi::new("/swagger-ui/{_:.*}").urls(vec![
                (
                   // we could have written the url as /api/v1/openapi.json as well for the first one and 
                   // then /api/v2/openapi.json for the second.
                   // but the urls provided to the SwaggerUi need to be relative from root of the server and it 
                   // does not care about the scope.
                    Url::new("api1", "/api-doc/openapi1.json"), // <- similarly is this url from root the server. 
                    ApiDoc1::openapi(),
                ),
            ]))
    })
    .bind((Ipv4Addr::UNSPECIFIED, 8080))?
    .run()
    .await
}

mod api1 {
    use actix_web::get;

    #[utoipa::path(
        context_path = "/api", // <- so we need to manually add the context path to the handlers behind the scope.
        responses(
            (status = 200, description = "Hello from api 1", body = String)
        )
    )]
    #[get("/api1/hello")]
    pub(super) async fn hello1() -> String {
        "hello from api 1".to_string()
    }
}
```
Hope this helps.

**Reply by tobihans** [reply]
Thank you for you answer. This is the best approach.
